### PR TITLE
 Allows mac shortcuts to be used when inside a input element

### LINKED ISSUE #12278: Support for native keybindings on input fields
This has being one of the issues that ppl raised on hacker news, in response to nightlies announcement. Most native keybindings (at least on Mac) do not work on input fields in servo. Example being

Alt+LeftArrow / Alt+RightArrow - Jump to previous / next word in the input.
Control+Alt+B / Control+Alt+F - Jump to previous / next word in the input.
CMD+LeftArrow / CMD+RightArrow - Jump to beginning / end of the input.
Control+Alt+A / Control+Alt+E - Jump to the beginning / end of the input.

All of the above with `+Shift` doing same + modifying selection.

### PR DESCRIPTION
**Pull request by jimberlage**
This adds the following navigation helpers on macOS:

Alt+LeftArrow / Alt+RightArrow - Jump to previous / next word in the
input.
Control+Alt+B / Control+Alt+F - Jump to previous / next word in the
input.
CMD+LeftArrow / CMD+RightArrow - Jump to beginning / end of the input.
Control+A / Control+E - Jump to the beginning / end of the
input.

It also respects shift when using these shortcuts to make a selection.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #12278 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12549)

### CODE REVIEW

**Comment by paulrouget** [comment]
Thank you for the PR.

Maybe in the future we'll want to use widget code for keybindings. For example: https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/EventOverview/TextDefaultsBindings/TextDefaultsBindings.html

**Comment by paulrouget** [comment]
I can review the behavior, but I'd like someone else to review the rust code (It looks good to me though).

Cmd+up/down, Ctrl+up/down, Cmd+home/end, Ctrl+a/e should move the cursor at the beginning/end of the whole text.

Option+up/down, should move the cursor at the beginning/end of the current line or of the next line if at the beginning/end of the current line. 

There's probably more moves I missed. I think we should cover the bindings found in Gecko:
- generic: https://dxr.mozilla.org/mozilla-central/source/dom/xbl/builtin/textareas-base.inc
- mac specific: https://dxr.mozilla.org/mozilla-central/source/dom/xbl/builtin/mac/platformHTMLBindings.xml#33
- I'm wondering if we also use this: https://dxr.mozilla.org/mozilla-central/source/dom/xbl/builtin/emacs/platformHTMLBindings.xml#14

We don't have to support the editing related bindings in this PR I think. Let's focus on getting the moves right.

File: components/script/textinput.rs [hunk 1]
```diff
@@ -393,6 +393,148 @@ impl<T: ClipboardProvider> TextInput<T> {
...
+            let current_line = &self.lines[self.edit_point.line];
+            match direction {
+                Direction::Forward => {
+                    let mut i = 0;
+
+                    // current_line[self.edit_point.index..].length() won't cut it here, as each
+                    // UTF-8 character has variable size.
+                    for ch in current_line[self.edit_point.index..].chars() {
+                        i = i + (ch.len_utf8() as isize);
+                    }
```
Comment (pcwalton) [inline components/script/textinput.rs]: How about `current_line[self.edit_point.index..].chars().map(char::len_utf8()).sum()`?

File: components/script/textinput.rs [hunk 2]
```diff
@@ -393,6 +393,148 @@ impl<T: ClipboardProvider> TextInput<T> {
...
+                Direction::Backward => {
+                    let mut i = 0;
+                    let mut iter = current_line[..self.edit_point.index].chars();
+
+                    loop {
+                        match iter.next_back() {
+                            Some(ch) => i = i - (ch.len_utf8() as isize),
+                            None => break,
+                        }
+                    }
```
Comment (pcwalton) [inline components/script/textinput.rs]: How about `i -= current_line[..self.edit_point.index].chars().map(char::len_utf8()).sum() as isize`?

File: components/script/textinput.rs [hunk 3]
```diff
@@ -480,16 +622,36 @@ impl<T: ClipboardProvider> TextInput<T> {
...
+                KeyReaction::RedrawSelection
+            },
             (Some('c'), _) if is_control_key(mods) => {
                 if let Some(text) = self.get_selection_text() {
                     self.clipboard_provider.set_clipboard_contents(text);
                 }
                 KeyReaction::DispatchInput
             },
+            #[cfg(target_os = "macos")]
+            (Some('e'), _) if (mods == CONTROL) || (mods == CONTROL | SHIFT) => {
```
Comment (pcwalton) [inline components/script/textinput.rs]: Some of these parentheses are unnecessary.

File: components/script/textinput.rs [hunk 4]
```diff
@@ -507,14 +669,34 @@ impl<T: ClipboardProvider> TextInput<T> {
...
+            #[cfg(target_os = "macos")]
+            (None, Key::Right) if is_control_key(mods) => {
+                self.adjust_horizontal_to_end(Direction::Forward, maybe_select);
+                KeyReaction::RedrawSelection
+            },
             (None, Key::Right) => {
                 self.adjust_horizontal_by_one(Direction::Forward, maybe_select);
                 KeyReaction::RedrawSelection
-            }
+            },
```
Comment (pcwalton) [inline components/script/textinput.rs]: nit: Remove this comma (and likewise for the above pattern arms)

File: components/script/textinput.rs [hunk 5]
```diff
@@ -393,6 +393,148 @@ impl<T: ClipboardProvider> TextInput<T> {
...
+    ///
+    /// "he<cursor>llo world\n\nhi" would be 3, since the cursor must move to the end of "hello".
+    /// "hello<cursor> world\n\nhi" would be 6, since the cursor must move to the end of "world".
+    /// "hello world<cursor>\n\nhi" would be 4, since the cursor must move to the end of "hi".
+    fn word_end_adjustment(&self) -> isize {
+        let mut adjust = 0;
+        let mut found_non_whitespace = false;
+        let mut edit_point_index = Some(self.edit_point.index);
+
+        'line: for line in self.lines[self.edit_point.line..].iter() {
```
Comment (pcwalton) [inline components/script/textinput.rs]: nit: `for line in &self.lines[self.edit_point.line..] {`

File: components/script/textinput.rs [hunk 6]
```diff
@@ -393,6 +393,148 @@ impl<T: ClipboardProvider> TextInput<T> {
...
+            let char_iter = match edit_point_index {
+                Some(i) => line[i..].chars(),
+                None => line.chars(),
+            };
+
+            'chars: for ch in char_iter {
```
Comment (pcwalton) [inline components/script/textinput.rs]: Could just be `for ch in line[edit_point_index.unwrap_or(0)..].chars() {`. Also no need for the `'chars:`

File: components/script/textinput.rs [hunk 7]
```diff
@@ -393,6 +393,148 @@ impl<T: ClipboardProvider> TextInput<T> {
...
+
+                adjust = adjust + (ch.len_utf8() as isize);
+            }
+
+            // If we've found some part of a word, we can stop at the end of the
+            // current line.  Otherwise, we need to jump to the end of the first word
+            // of the next line.
+            if found_non_whitespace { break 'line; }
+
+            adjust = adjust + 1; // Account for newline character
```
Comment (pcwalton) [inline components/script/textinput.rs]: `adjust += 1;`

**Comment by jdm** [comment]
Closing due to lack of activity.

**Pull request closed without merging**